8282648: Weaken the InflaterInputStream specification in order to allow faster Zip implementations

simonis opened the pull request:
Add an API note to `InflaterInputStream::read(byte[] b, int off, int len)` to highlight that it might  write more bytes than the returned number of inflated bytes into the buffer `b`.

The superclass `java.io.InputStream` specifies that `read(byte[] b, int off, int len)` will leave the content beyond the last read byte in the read buffer `b` unaffected. However, the overridden `read` method in `InflaterInputStream` passes the read buffer `b` to `Inflater::inflate(byte[] b, int off, int len)` which doesn't provide this guarantee. Depending on implementation details, `Inflater::inflate` might write more than the returned number of inflated bytes into the buffer `b`.

### TL;DR

`java.util.zip.Inflater` is the Java wrapper class for zlib's inflater functionality. `Inflater::inflate(byte[] output, int off, int len)` currently calls zlib's native `inflate(..)` function and passes the address of `output[off]` and `len` to it via JNI.

The specification of zlib's `inflate(..)` function (i.e. the [API documentation in the original zlib implementation](https://github.com/madler/zlib/blob/cacf7f1d4e3d44d871b605da3b647f07d718623f/zlib.h#L400)) doesn't give any guarantees with regard to usage of the output buffer. It only states that upon completion the function will return the number of bytes that have been written (i.e. "inflated") into the output buffer.

The original zlib implementation only wrote as many bytes into the output buffer as it inflated. However, this is not a hard requirement and newer, more performant implementations of the zlib library like [zlib-chromium](https://chromium.googlesource.com/chromium/src/third_party/zlib/) or [zlib-cloudflare](https://github.com/cloudflare/zlib) can use more bytes of the output buffer than they actually inflate as a scratch buffer. See https://github.com/simonis/zlib-chromium for a more detailed description of their approach and its performance benefit.

These new zlib versions can still be used transparently from Java (e.g. by putting them into the `LD_LIBRARY_PATH` or by using `LD_PRELOAD`), because they still fully comply to specification of `Inflater::inflate(..)`. However, we might run into problems when using the `Inflater` functionality from the `InflaterInputStream` class. `InflaterInputStream` is derived from from `InputStream` and as such, its `read(byte[] b, int off, int len)` method is quite constrained. It specifically specifies that if *k* bytes have been read, then "these bytes will be stored in elements `b[off]` through `b[off+`*k*`-1]`, leaving elements `b[off+`*k*`]` through `b[off+len-1]` **unaffected**". But `InflaterInputStream::read(byte[] b, int off, int len)` (which is constrained by `InputStream::read(..)`'s specification) calls `Inflater::inflate(byte[] b, int off, int len)` and directly passes its output buffer down to the native zlib `inflate(..)` method which is free to change the bytes beyond `b[off+`*k*`]` (where *k* is the number of inflated bytes).

From a practical point of view, I don't see this as a big problem, because callers of `InflaterInputStream::read(byte[] b, int off, int len)` can never know how many bytes will be written into the output buffer `b` (and in fact its content can always be completely overwritten). It therefore makes no sense to depend on any data there being untouched after the call. Also, having used zlib-cloudflare productively for about two years, we haven't seen real-world issues because of this behavior yet. However, from a specification point of view it is easy to artificially construct a program which violates `InflaterInputStream::read(..)`'s postcondition if using one of the alterantive zlib implementations. A recently integrated JTreg test (test/jdk/jdk/nio/zipfs/ZipFSOutputStreamTest.java) "unintentionally" fails with zlib-chromium but can fixed easily to run with alternative implementations as well (see [JDK-8283756](https://bugs.openjdk.java.net/browse/JDK-8283756)).

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires a CSR request to be approved

### Issues
 * [JDK-8282648](https://bugs.openjdk.org/browse/JDK-8282648): Weaken the InflaterInputStream specification in order to allow faster Zip implementations
 * [JDK-8283758](https://bugs.openjdk.org/browse/JDK-8283758): Weaken the InflaterInputStream specification in order to allow faster Zip implementations (**CSR**)


### Reviewers
 * [Lance Andersen](https://openjdk.org/census#lancea) (@LanceAndersen - **Reviewer**)
 * [Alan Bateman](https://openjdk.org/census#alanb) (@AlanBateman - **Reviewer**) ⚠️ Review applies to [d82c7526](https://git.openjdk.org/jdk/pull/7986/files/d82c75267a76d3538465de17f4f7ffd2962dd563)
 * [Jaikiran Pai](https://openjdk.org/census#jpai) (@jaikiran - **Reviewer**) ⚠️ Review applies to [c9b60404](https://git.openjdk.org/jdk/pull/7986/files/c9b60404b39241d46ec32c86d35839a9f7a9659d)
 * [Mark Reinhold](https://openjdk.org/census#mr) (@mbreinhold - **Reviewer**) ⚠️ Review applies to [d82c7526](https://git.openjdk.org/jdk/pull/7986/files/d82c75267a76d3538465de17f4f7ffd2962dd563)
 * [Joe Darcy](https://openjdk.org/census#darcy) (@jddarcy - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/7986/head:pull/7986` \
`$ git checkout pull/7986`

Update a local copy of the PR: \
`$ git checkout pull/7986` \
`$ git pull https://git.openjdk.org/jdk pull/7986/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 7986`

View PR using the GUI difftool: \
`$ git pr show -t 7986`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/7986.diff">https://git.openjdk.org/jdk/pull/7986.diff</a>

</details>

simonis commented:
/csr

simonis commented:
> Hi Volker,
> 
> I believe your PR should point to the [JBS issue](https://bugs.openjdk.java.net/browse/JDK-8282648) in the title, which references the CSR and not the CSR directly in the title.

Sorry, you're right of course :)

jaikiran commented on src/java.base/share/classes/java/util/zip/InflaterInputStream.java:
```diff
@@ -127,10 +127,15 @@ public int read() throws IOException {
+     *
+     * Unlike the {@link InputStream#read(byte[],int,int) overridden method}
+     * of {@code InputStream}, this method might write more bytes than the returned
+     * number of inflated bytes into the buffer {@code b}.
```
Hello Volker, I think this comment should be a bit more clear and state what exactly it means by writing more bytes than the returned value. Specifically, I think it should clearly state that even though it might write additional data, it will however not write/change data starting and beyond `off + len` index in the passed array.

AlanBateman replied:
I think this require a bit of word smithing. If the return value is 'n' then it should make it clear that the values in elements b[off + n] to b[off + len - 1] are undefined, their values may or may have been changed by the inflate/read operation. For completeness, the expectation for when inflate fails should be specified too, the contents of b[off] to b[off + len - 1] will be undefined.

simonis replied:
Thanks for your comments. I've updated the API-doc according to your recommendations. Please feel free to propose new/better wording :)

I'll wait with updating the CSR until we reach final agreement in this PR.

jaikiran commented:
Hello Volker,
An additional thing that we might have to consider here is whether adding this javadoc change to `InflaterInputStream` is ever going to "show up" to end user applications.
What I mean is, I think in many cases the end user applications won't even know they are dealing with an `InflaterInputStream`. For example, the following code:

```
ZipFile zf = ...
ZipEntry ze = zf.getEntry("some-file");
InputStream is = zf.getInputStream(ze);
```
As we see above, none of these APIs talk about `InflaterInputStream` (the return type of `ZipFile.getInpustream(...)` is an `InputStream`). So end users won't be able to view this spec change. Perhaps we should also add some note in the `ZipFile.getInpustream(....)` API to make a mention of this potential difference in behaviour of the returned stream?

simonis commented:
> Hello Volker, An additional thing that we might have to consider here is whether adding this javadoc change to `InflaterInputStream` is ever going to "show up" to end user applications. What I mean is, I think in many cases the end user applications won't even know they are dealing with an `InflaterInputStream`. For example, the following code:
> 
> ```
> ZipFile zf = ...
> ZipEntry ze = zf.getEntry("some-file");
> InputStream is = zf.getInputStream(ze);
> ```
> 
> As we see above, none of these APIs talk about `InflaterInputStream` (the return type of `ZipFile.getInpustream(...)` is an `InputStream`). So end users won't be able to view this spec change. Perhaps we should also add some note in the `ZipFile.getInpustream(....)` API to make a mention of this potential difference in behaviour of the returned stream?

You are right with your observation and I'll be happy to add a corresponding comment if @LanceAndersen and @AlanBateman agree. Please let me know what you think?

RogerRiggs commented:
One other way to communicate changes is in the release-note.  I added release-note=yes to the bug.

LanceAndersen commented:
> > Hello Volker, An additional thing that we might have to consider here is whether adding this javadoc change to `InflaterInputStream` is ever going to "show up" to end user applications. What I mean is, I think in many cases the end user applications won't even know they are dealing with an `InflaterInputStream`. For example, the following code:
> > ```
> > ZipFile zf = ...
> > ZipEntry ze = zf.getEntry("some-file");
> > InputStream is = zf.getInputStream(ze);
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > As we see above, none of these APIs talk about `InflaterInputStream` (the return type of `ZipFile.getInpustream(...)` is an `InputStream`). So end users won't be able to view this spec change. Perhaps we should also add some note in the `ZipFile.getInpustream(....)` API to make a mention of this potential difference in behaviour of the returned stream?
> 
> You are right with your observation and I'll be happy to add a corresponding comment if @LanceAndersen and @AlanBateman agree. Please let me know what you think?

Hi Volker,

I believe Jai raises a valid point given these javadocs probably have had limited updates if any since the API was originally added.    We should look at ZipInputStream and GZipInputStream as well if we decide to update the ZipFile::getInputStream(where we could borrow some wording from the ZipInputStream class description as a start to some word smithing).

As Roger points out we will need a release note for this change as well.

AlanBateman commented:
A suggestion for the structure is to start the new paragraph by saying it reads n bytes of uncompressed data into b[off] to b[off+n-1]. Then follow to say that the contents of b[n] to b[off+len-1] after the read are are undefined, then say that an implementation is free to ...   Finally just finish it by saying that the contents are also undefined when the method fails by throwing an exception.

simonis commented:
Hi,

I've pushed a new version which:
- improves the wording as suggested
- adds the API amendments from `InflaterInputStream::read()` to `GZIPInputStream::read()`/`ZipInputStream::read()` as well be cause they both inherit from `InflaterInputStream`. Unfortunately it's not possible to simply inherit the API doc because of different parameter naming.
- added an `@implNote` to `ZipFile::getInputStream()` to make it clear that this implementation really returns an `InflaterInputStream`.

Please let me know what you think?
Volker

jaikiran commented on src/java.base/share/classes/java/util/zip/GZIPInputStream.java:
```diff
@@ -95,6 +95,14 @@ public GZIPInputStream(InputStream in) throws IOException {
      * no bytes are read and {@code 0} is returned.
+     * <p>
+     * If <i>n</i> denotes the returned number of inflated bytes then {@code buf[off]}
+     * trough {@code buf[off+}<i>n</i>{@code -1]} will contain the uncompressed data. The
```
Hello Volker, there's a typo here - "trough" should be "through"

jaikiran commented on src/java.base/share/classes/java/util/zip/GZIPInputStream.java:
```diff
@@ -95,6 +95,14 @@ public GZIPInputStream(InputStream in) throws IOException {
+     * content of {@code buf} is undefined.
+     *
      * @param buf the buffer into which the data is read
      * @param off the start offset in the destination array {@code b}
```
Although not related to this PR, since we are already in this area, could you also please fix this and change `{@code b}` to `{@code buf}`

jaikiran commented on src/java.base/share/classes/java/util/zip/ZipFile.java:
```diff
@@ -343,6 +343,9 @@ public ZipEntry getEntry(String name) {
      * Closing this ZIP file will, in turn, close all input streams that
      * have been returned by invocations of this method.
      *
+     * @implNote In the JDK implementation this method returns an
```
The "In the JDK implementation" sounds a bit odd to me. I looked around some other `@implNote`s and many of of them use `@implNote This implementation ....`, so perhaps we should change this to:
``` 
@implNote This implementation returns an instance of {@link java.util.zip.InflaterInputStream}.
```

simonis replied:
Thanks for the suggestion. I've updated to code accordingly.

jaikiran commented on src/java.base/share/classes/java/util/zip/ZipInputStream.java:
```diff
@@ -169,6 +169,14 @@ public int available() throws IOException {
      * blocks until some input is available; otherwise, no
      * bytes are read and {@code 0} is returned.
+     * <p>
+     * If <i>n</i> denotes the returned number of inflated bytes then {@code b[off]}
```
I think this might need a slightly different wording since in this case of `ZipInputStream#read(...)`, the entry's  compression method decides whether or not an inflater will be used to return inflated bytes.

So maybe we should just remove the reference to "inflated bytes" and instead say:
```
* If <i>n</i> denotes the return value, then {@code b[off]} through {@code b[off+}<i>n</i>{@code -1]}
* will contain the data that has been read. The elements {@code b[off+}<i>n</i>{@code ]} through 
* {@code b[off+}<i>len</i>{@code -1]} are undefined and the implementation may update them during this read
* operation. If the return value is -1 or an exception is thrown the whole content of {@code b} is undefined.
```

simonis replied:
Thanks for spotting this. As the restriction only applies to compressed entries, I've simply prefixed the text with "If the current entry is compressed and.."

jaikiran commented:
The `GZIPInputStream` and `ZipInputStream` files will need a copyright year update.

AlanBateman commented on src/java.base/share/classes/java/util/zip/ZipFile.java:
```diff
@@ -343,6 +343,9 @@ public ZipEntry getEntry(String name) {
      * have been returned by invocations of this method.
      *
+     * @implNote This implementation returns an instance of
+     * {@link java.util.zip.InflaterInputStream}.
```
What is the reasoning for this note, do we really want it in the API docs?

jaikiran replied:
Hello Alan, this change was done by Volker for the point that I raised in the comment here https://github.com/openjdk/jdk/pull/7986#issuecomment-1081319691

AlanBateman replied:
> Hello Alan, this change was done by Volker for the point that I raised in the comment here [#7986 (comment)](https://github.com/openjdk/jdk/pull/7986#issuecomment-1081319691)

I think it creates the temptation to cast the returned input stream to InflaterInputStream, it might be better to leave it out.

AlanBateman commented on src/java.base/share/classes/java/util/zip/GZIPInputStream.java:
```diff
@@ -95,8 +95,16 @@ public GZIPInputStream(InputStream in) throws IOException {
+     * @param off the start offset in the destination array {@code buf}
      * @param len the maximum number of bytes read
      * @return  the actual number of bytes read, or -1 if the end of the
      *          compressed input stream is reached
```
We should probably adjust the `@return` description, as was done for InflaterInputStream, otherwise the method. description will say "the returning number of inflated bytes" and the return description will say "the actual number of bytes read".

simonis replied:
Fixed (also in `ZipInputStream`).

AlanBateman commented on src/java.base/share/classes/java/util/zip/GZIPInputStream.java:
```diff
@@ -95,8 +95,16 @@ public GZIPInputStream(InputStream in) throws IOException {
+     * elements {@code buf[off+}<i>n</i>{@code ]} through {@code buf[off+}<i>len</i>{@code -1]}
+     * are undefined and an implementation is free to change them during the inflate
+     * operation. If the return value is -1 or an exception is thrown the whole
+     * content of {@code buf} is undefined.
```
"whole content" doesn't seem right, it should be buf[off] to but[off + len - 1] is undefined.

AlanBateman commented on src/java.base/share/classes/java/util/zip/GZIPInputStream.java:
```diff
@@ -95,8 +95,16 @@ public GZIPInputStream(InputStream in) throws IOException {
+     * through {@code buf[off+}<i>n</i>{@code -1]} will contain the uncompressed data. The
+     * elements {@code buf[off+}<i>n</i>{@code ]} through {@code buf[off+}<i>len</i>{@code -1]}
+     * are undefined and an implementation is free to change them during the inflate
+     * operation. If the return value is -1 or an exception is thrown the whole
```
The sentence is very long. How about putting "an implementation is free ..." in parentheses after "undefined".

LanceAndersen commented on src/java.base/share/classes/java/util/zip/GZIPInputStream.java:
```diff
@@ -95,10 +95,18 @@ public GZIPInputStream(InputStream in) throws IOException {
+     * through {@code buf[off+}<i>n</i>{@code -1]} will contain the uncompressed data. The
+     * elements {@code buf[off+}<i>n</i>{@code ]} through {@code buf[off+}<i>len</i>{@code -1]}
+     * are undefined (an implementation is free to change them during the inflate
+     * operation). If the return value is -1 or an exception is thrown then {@code buf[off]}
```
perhaps tweak "... is thrown then" 

to

"... is thrown, then the contents of..."

simonis replied:
Updated as requested (also in `ZipInputStream` and `InflaterInputStream`).

LanceAndersen commented on src/java.base/share/classes/java/util/zip/GZIPInputStream.java:
```diff
@@ -95,10 +95,18 @@ public GZIPInputStream(InputStream in) throws IOException {
      * zero, the method will block until some input can be decompressed; otherwise,
      * no bytes are read and {@code 0} is returned.
+     * <p>
+     * If <i>n</i> denotes the returned number of inflated bytes then {@code buf[off]}
```
I think a comma might be needed after "...inflated bytes"

LanceAndersen reviewed (commented):
Hi Volker,

I think this reads much better.  Its too bad we cannot take advantage of `@inheritedDoc`

A couple of minor comments below

AlanBateman reviewed (approved):
You've addressed my points, the updated javadoc looks good.

LanceAndersen reviewed (approved):
Hi Volker,

The updates look good.  Thank you.   Next up is updating the CSR with the changes and craft the release-note as a subtask.

Thank you for your efforts here

simonis commented:
Thanks @AlanBateman, @LanceAndersen. I've just updated the CSR with the latest wording from this PR. Please feel free to review the CSR as well so I can finalize it.

simonis commented:
@jddarcy requested in the [CSR](https://bugs.openjdk.java.net/browse/JDK-8283758) to directly update the API spec of `InputStream`. This considerably simplifies the whole patch because the derived classes now inherit the correct specification right from their base class and don't need to be updated anymore.

Please comment right on the [CSR](https://bugs.openjdk.java.net/browse/JDK-8283758) should you have any objections or suggestions.

simonis commented:
@LanceAndersen, @AlanBateman can you please comment on the [CSR](https://bugs.openjdk.java.net/browse/JDK-8283758) for this issue. We now circled back to the initial proposal in the  [CSR](https://bugs.openjdk.java.net/browse/JDK-8283758) but  @jddarcy would like to hear your opinion.

LanceAndersen commented:
> @LanceAndersen, @AlanBateman can you please comment on the [CSR](https://bugs.openjdk.java.net/browse/JDK-8283758) for this issue. We now circled back to the initial proposal in the [CSR](https://bugs.openjdk.java.net/browse/JDK-8283758) but @jddarcy would like to hear your opinion.

Have not forgotten about this.  I think we are not quite there on the changes to` Inputstream` but have not had a chance to give it some more thought.  Apologies and thank you for your patience

mbreinhold commented:
First, I’ll observe that there is no specification conflict here.  The specifications of `InflaterInputStream::read(byte[] b, int off, int len)` and `Inflater::inflate(byte[] b, int off, int len)` are not in conflict at all.  The latter is simply weaker than the former.

What is true is that if you build the JDK in the default manner, without any patches, then what you wind up with is an `InflaterInputStream::read(...)` implementation that satisfies the inherited `InputStream::read(...)` specification, including the constraint that a read operation must not _scribble_ in the caller’s output buffer, _i.e._, modify any byte in that buffer beyond the returned byte count.  (That constraint is, so far as we know, satisfied by every `InputStream` subclass defined in the JDK.)

A default JDK build’s `InflaterInputStream::read(...)` method doesn’t scribble because

  - The implementation of `InflaterInputStream::read(...)` assumes that the implementation of `Inflater::inflate(...)` doesn’t scribble, and

  - The implementation of `Inflater::inflate(...)` satisfies that assumption because it invokes the corresponding function in the [standard native Zip implementation](https://zlib.net/), which itself doesn’t scribble.

So, not only does the specification forbid `InflaterInputStream::read(...)` from scribbling, but in the official Java SE Reference Implementation — which is simply a default JDK build — that method does not scribble.  (Perhaps by accident, but that is immaterial here.)

Therefore, if in your own JDK build you substitute an alternative Zip implementation whose `inflate` function does scribble then you wind up with a JDK that both violates the specification and behaves differently than the RI.  It is incompatible.  If you use such a JDK in production then you risk breaking existing code.

Now, compatibility is one of the key values of the Java Platform, but performance is not unimportant.  Is there something we can do to accommodate [alternative, faster, scribbling Zip implementations][faster-zlib]?

@simonis lists three options in the [issue].  I’ll cast the options somewhat differently.

  1. Do not change the specification in any way.  The specification text that forbids scribbling dates back to Java 1.1, released over 25 years ago.  (Arguably that text over-specifies the `InputStream::read(...)` method, but that is also immaterial.)  It’s all too possible that existing code — no matter how odd or poorly written we think it might be — depends upon this constraint.

     With this option, a JDK build that uses a scribbling Zip implementation would have to buffer that implementation’s output and copy only the returned number of bytes into the caller’s buffer.  This would degrade any performance benefit.

  2. Weaken the specification of `InflaterInputStream` — and, thereby, that of its subclasses `GZipInputStream` and `ZipInputStream` — to allow scribbling.  This is essentially @simonis’s initial suggestion, as embodied in [63ae357] earlier in this PR.

     This option risks breaking existing code that depends upon the 25-year-old promise that `InflaterInputStream::read(...)` will not scribble.  Changing the specification of an indirect subclass of `InputStream`, moreover, is apt to go unnoticed by many developers, who routinely deal with instances of `InputStream` without any knowledge of those instances’ concrete classes.  (@jaikiran gives one example above.)

  3. Weaken the specification of the `InputStream::read(...)` method to allow scribbling, as [suggested by @jddarcy in the CSR][jddarcy].

     This option is at least as risky to existing code as the second option, though it’s marginally more likely that developers writing new code will notice the specification change.

     What’s worse, however, is that this option gives developers writing new subclasses of `InputStream` permission to deviate from both that class’s longstanding specification and the actual behavior of every `InputStream` subclass defined in the JDK.  In the long term, that could wreak subtle bugs throughout the ecosystem, even breaking code that does absolutely nothing with Zip-related input streams.

In favor of both the second and third options is @simonis’s statement above that his employer (Amazon) has used JDK builds with a scribbling Zip implementation in production for two years.  He has further elaborated (privately) that this includes hundreds of services built with many of the usual popular frameworks and libraries (Spring, Hibernate, ASM, etc.).  This suggests that the practical impact, out in the wild, of allowing scribbling Zip implementations is acceptable.

On that basis I am — just barely — comfortable with the second option, _i.e._, weakening the specification of `InflaterInputStream` to allow scribbling, thereby contradicting the specification of its `InputStream` superclass.

Explicitly specifying a subclass in a way that’s inconsistent with the specification of one of its superclasses is generally a really bad idea.  It violates the [Liskov Substitution Principle][lsp], which enables instances of a subclass to be substituted freely for instances of one of its superclasses.  Thus we do not condone such changes lightly.  There are, however, precedents in a handful of special cases in the Java Platform; _e.g._, in the [`java.util.IdentityHashMap`][ihm] and [`java.sql.Timestamp`][ts] classes.

The third option above would not violate the Liskov substitution principle, but in my view could prove more harmful in the long run.

Recommendations:

  - Revert this PR back to [63ae357], and make corresponding updates to the [issue] and the [CSR].

  - Additionally, find every method in every `java.*` class that’s declared to return an `InputStream` (_e.g._, `ZipFile::getInputStream(...)` but actually returns an instance of `InflaterInputStream` or one of its subclasses.  Add an API note to each such method to let developers know that the returned input stream might scribble, per @jaikiran’s suggestion above.  These notes can all use the same text, linking to the revised `InflaterInputStream::read(...)` specification.

  - Revise the summary of this PR, the issue, and the CSR to reflect the true nature of the proposed change rather than assert a nonexistent conflict.  Consider “Weaken the `InflaterInputStream` specification in order to allow faster Zip implementations”.

  - Change the issue type from a bug — which it’s not — to an enhancement.

  - Create a related release-note issue, as @RogerRiggs suggests above.


[faster-zlib]: https://github.com/simonis/zlib-chromium#inflater-improvements-in-zlib-chromium
[issue]: https://bugs.openjdk.org/browse/JDK-8282648
[63ae357]: https://github.com/openjdk/jdk/pull/7986/commits/63ae3572da674a0926bb9d0adcea88d89bb56707
[jddarcy]: https://bugs.openjdk.org/browse/JDK-8283758?focusedCommentId=14492930&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-14492930
[CSR]: https://bugs.openjdk.org/browse/JDK-8283758
[lsp]: https://en.wikipedia.org/wiki/Liskov_substitution_principle
[ihm]: https://docs.oracle.com/en/java/javase/18/docs/api/java.base/java/util/IdentityHashMap.html
[ts]: https://docs.oracle.com/en/java/javase/18/docs/api/java.sql/java/sql/Timestamp.html

simonis commented:
The latest push contains the following changes:
- changed the summary of the Issue and CSR to "Weaken the InflaterInputStream specification in order to allow faster Zip implementations" as suggested
- rebased the PR up to change [63ae357](https://github.com/openjdk/jdk/pull/7986/commits/63ae3572da674a0926bb9d0adcea88d89bb56707) on top of the leatest HEAD revision
- Applied the same changes to the JavaDoc of `JarInputStream::read()` like to ones which were added to `ZipInputStream::read()`. `JarInputStream` which derives from `ZipInputStream` and overrides `ZipInputStream::read()` was just forgotten in the previous itetrations of this change.
- Added an `@apiNote` to `ZipFile::getInputStream()`, `JarFile::getInputStream()` and `URLConnection::getInputStream()` to indicate that the returned `InputStream` might *scribble* the contents of the output array beyond the last inflated byte. I couldn't find any other public methods in the `java.*` packages which return `InflaterInputStream` instances (or derived subclasses) as a generic `InputStream`.

mbreinhold commented on src/java.base/share/classes/java/util/zip/InflaterInputStream.java:
```diff
@@ -124,13 +124,21 @@ public int read() throws IOException {
+     * elements {@code b[off+}<i>n</i>{@code ]} through {@code b[off+}<i>len</i>{@code -1]}
+     * are undefined (an implementation is free to change them during the inflate
+     * operation). If the return value is -1 or an exception is thrown, then the content of
+     * {@code b[off]} to {@code b[off+}<i>len</i>{@code -1]} is undefined.
```
Consider this more precise wording, both here and in `GZIPInputStream`, `ZipInputStream`, and `JarInputStream`:
```
     * If this method returns a nonzero integer <i>n</i> then {@code buf[off]}
     * through {@code buf[off+}<i>n</i>{@code -1]} contain the uncompressed
     * data.  The content of elements {@code buf[off+}<i>n</i>{@code ]} through
     * {@code buf[off+}<i>len</i>{@code -1]} is undefined, contrary to the
     * specification of the {@link java.io.InputStream InputStream} superclass,
     * so an implementation is free to modify these elements during the inflate
     * operation. If this method returns {@code -1} or throws an exception then
     * the content of {@code buf[off]} through {@code buf[off+}<i>len</i>{@code
     * -1]} is undefined.
```

mbreinhold commented on src/java.base/share/classes/java/util/zip/InflaterInputStream.java:
```diff
@@ -124,13 +124,21 @@ public int read() throws IOException {
      * @param off the start offset in the destination array {@code b}
      * @param len the maximum number of bytes read
-     * @return the actual number of bytes read, or -1 if the end of the
+     * @return the actual number of inflated bytes, or -1 if the end of the
```
s/inflated bytes/bytes inflated/

mbreinhold commented on src/java.base/share/classes/java/net/URLConnection.java:
```diff
@@ -840,6 +840,13 @@ public Permission getPermission() throws IOException {
+     * which leaves the contents of the output buffer beyond the
+     * last inflated byte undefined after a read operation (see {@link
+     * java.util.zip.InflaterInputStream#read(byte[], int, int)
+     * InflaterInputStream.read(byte[], int, int)}).
```
Consider this wording, which makes the danger clearer, here and elsewhere:
```
     * @apiNote The {@code InputStream} returned by this method can wrap an
     * {@link java.util.zip.InflaterInputStream InflaterInputStream}, whose
     * {@link java.util.zip.InflaterInputStream#read(byte[], int, int)
     * read(byte[], int, int)} method can modify any element of the output
     * buffer.
```

mbreinhold reviewed (changes requested):
Please update the CSR once you’ve finalized the specification changes.

LanceAndersen reviewed (approved):
Volker,

Thank you for your patience while we worked through all of the nuances of this 

The changes look good to me

simonis commented:
@mbreinhold, @LanceAndersen thanks for your reviews. I'm waiting for the CSR approval before pushing.

AlanBateman reviewed (approved):
The update in d82c7526 looks good.

jddarcy commented on src/java.base/share/classes/java/util/jar/JarInputStream.java:
```diff
@@ -167,10 +167,21 @@ public JarEntry getNextJarEntry() throws IOException {
-     * If {@code len} is not zero, the method
-     * blocks until some input is available; otherwise, no
-     * bytes are read and {@code 0} is returned.
+     * Reads from the current ZIP entry into an array of bytes, returning the number of
```
Please change "Reads from the current ZIP entry..." to "Reads from the current JAR entry..."

simonis replied:
Thanks for your review and for spotting this copy&paste error. I've fixed it.

LanceAndersen reviewed (approved):
Typo cleanup looks fine

simonis commented:
/integrate
